### PR TITLE
Add script to set authorities. Convert to object in Ruby to manage auth/proxy context.

### PR DESCRIPTION
**Pull request by markfickett**
What do you think of this approach for handling auth and CloudSQL proxy setup/teardown? To me a collection of related things you have to do before running and then after running => an object to encapsulate those things. I was trying to do it just with Ruby blocks, but couldn't get it to gel.

This implementation isn't done -- I still need to make the do_run_with_creds part happen for my new GcloudContext -- but it's close enough to demonstrate the approach.

Upside is that I can do:

```
workbench$ api/project.rb set-authority --help
Usage: ./project.rb set-authority [options]
        --project [PROJECT]          Project to create credentials for (e.g. all-of-us-workbench-test)
        --account [ACCOUNT]          Account to use when creating credentials (your.name@pmi-ops.org); use this or --creds_file
        --creds_file [CREDS_FILE]    Path to a file containing credentials; use this or --account.
        --email [EMAIL,...]          Comma-separated list of user accounts to change. Required.
        --add_authority [AUTHORITY,...]
                                     Comma-separated list of user authorities to add for the users. One of added or removed authorities is required.
        --rm_authority [AUTHORITY,...]
                                     Comma-separated list of user authorities to remove from the users.

# missing a required arg, shows an error immediately
workbench$ api/project.rb set-authority --account mark.fickett@pmi-ops.org --project all-of-us-workbench-test --email mark.fickett@staging.pmi-ops.org

# validation succeeds, CloudSQL proxy set up, gradle build starts
workbench$ api/project.rb set-authority --account mark.fickett@pmi-ops.org --project all-of-us-workbench-test --email mark.fickett@staging.pmi-ops.org --add_authority REVIEW_RESEARCH_PURPOSE
Running Cloud SQL Proxy...
+ sleep 3
2017/10/13 15:59:11 google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
+ /usr/local/google/home/mwf/mwf/git/workbench/api/gradlew --info setAuthority -PappArgs="['mark.fickett@staging.pmi-ops.org','REVIEW_RESEARCH_PURPOSE','']"
...
```

### CODE REVIEW

File: api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+  @Bean
+  public CommandLineRunner run(UserDao userDao) {
+    return (args) -> {
+      if (args.length != 3) {
+        throw new IllegalArgumentException(
+            "Expected email_list, authority_list_to_add, authority_list_to_rm. Got " + args);
+      }
+      Set<String> emails = new HashSet();
+      emails.add("mark.fickett@staging.pmi-ops.org");
```
Comment (danqrodney) [inline api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java]: This part is changing, right?

File: api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        if (user == null) {
+          log.log(Level.SEVERE, "No user for {1}.", email);
+          numErrors++;
+          continue;
+        }
+        Set<Authority> granted = user.getAuthorities();
+        granted.addAll(authoritiesToAdd);
+        granted.removeAll(authoritiesToRemove);
+        if (!dryRun) {
+          user.setAuthorities(granted);
```
Comment (danqrodney) [inline api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java]: I think you need userDao.save(user) here?

File: api/libproject/devstart.rb [hunk 3]
```diff
@@ -413,6 +412,94 @@ def get_test_service_account_creds(*args)
...
+    end
+  end
+
+  def run
+    super do
+      Dir.chdir("tools") do
+        gradlew_path = File.join(Workbench::WORKBENCH_ROOT, "api", "gradlew")
+        @common.run_inline %W{
+            #{gradlew_path} --info setAuthority
+            -PappArgs="['#{@opts.email}','#{@opts.add_authority}','#{@opts.rm_authority}']"}
```
Comment (danqrodney) [inline api/libproject/devstart.rb]: Does this separate them with spaces? What happens if one of them is missing?

File: api/libproject/devstart.rb [hunk 4]
```diff
@@ -253,11 +253,11 @@ def get_auth_login_account()
 
 def run_with_creds(args, command_name, proc)
```
Comment (danqrodney) [inline api/libproject/devstart.rb]: Is this being replaced by GcloudContext like below? (Are you planning on changing all other uses to do that?)

File: api/tools/build.gradle [hunk 5]
```diff
@@ -9,20 +9,31 @@ def db_port = System.getenv("DB_PORT")
...
 }
 
+// See api/project.rb set-authority.
+// This evals the value of the -PappArgs flag given to Gradle and passes the
+// result to the Java main method.
+task setAuthority(type: JavaExec) {
+  classpath sourceSets.main.runtimeClasspath
+  main = "org.pmiops.workbench.tools.SetAuthority"
+  systemProperties = dbProperties
+  args Eval.me(appArgs)
```
Comment (danqrodney) [inline api/tools/build.gradle]: This seems fine. Also fine would have been to list out the names of expected params here, as above with config_file.

File: api/libproject/devstart.rb (hunk 3 above)
Comment (danqrodney) [reply api/libproject/devstart.rb]: (In general, I'm a little confused about how this is going to get parsed by the Java code...)

File: api/libproject/devstart.rb (hunk 4 above)
Comment (markfickett) [reply api/libproject/devstart.rb]: Yes, I think I'd like to switch everything over so it's uniform, but wanted to just do a couple demo use cases to start.

File: api/libproject/devstart.rb (hunk 3 above)
Comment (markfickett) [reply api/libproject/devstart.rb]: The Java code would just get 3 positional args. Right now if one is missing you get an empty string, which I think will work out OK, but I haven't tried it yet.

I'm just looking for a way to front-load all the parsing/validation smarts, and keep everything in this particular function, and Gradle, and the Java, all as dumb as possible (meaning, requiring as little change as possible when you change the command-line flags).

File: api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java (hunk 1 above)
Comment (markfickett) [reply api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java]: Yes -- I just hardcoded some values to start trying out the Java before I figured out my command-line parsing strategy. I think what I'd do is something like `Set<String> emails = parseCommaSeparatedString(args[0]);`.

File: api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java (hunk 2 above)
Comment (markfickett) [reply api/tools/src/main/java/org/pmiops/workbench/tools/SetAuthority.java]: Thanks -- probably! This file doesn't even compile yet.

File: api/libproject/devstart.rb (hunk 3 above)
Comment (danqrodney) [reply api/libproject/devstart.rb]: OK... I just wasn't sure how you'd distinguish between add_authority and rm_authority when there are only two non-empty args passed in. Is there going to be some prefix or something?

Comment (danqrodney) [reply api/libproject/devstart.rb]: An alternative that would be less ambiguous is:

<op> <emails> <authorities>

where <op> is "add" or "rm".

I don't think you really need to support adding and removing at the same time. At the moment we only have one authority anyway. :)

**Comment by markfickett** [comment]
PTAL. Ruby is all changed over to the new style, and Java is working now.

Example command line:

`api/project.rb set-authority --project all-of-us-workbench-test --account mark.fickett@pmi-ops.org --email mark.fickett@staging.pmi-ops.org --add_authority REVIEW_RESEARCH_PURPOSE`

Ruby validates that and converts it to `appArgs=['mark.fickett@pmi-ops.org','REVIEW_RESEARCH_PURPOSE','', false]`, which Gradle `eval`s as a 4-element Java array (3 Strings and a boolean) passed in as `args` in `SetAuthority.run`. (The empty string just gets parsed as an empty `Set<Authority>` of permissions to remove, in this case.)

**Pull request merged**